promql/engine.go

```diff
@@ -1281,6 +1298,7 @@ func (ev *evaluator) eval(expr parser.Expr) parser.Value {
 			for ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval {
 				_, v, ok := ev.vectorSelectorSingle(it, e, ts)
 				if ok {
+					ev.samplesStats.Increment(1)
```

Comment:
I think we're double counting here, as we'll also count these when they're an input to rangeEval.

This means that a query that is only a selector will have a total of 0, as no math was done.

Reply:
Yes, we are double counting. That's what I said way before in comments.

Reply:
That's why I said that total samples should be only the count in vector_selector and not in any other places since vector selectors are the only area that fetches the samples/series.

Reply:
That'd under count though, it's the number of samples that are passed into rate that determine performance more than the number of samples fetched. A rate(1h) with a 1m step is more expensive than a rate 1h with a 1h step over the same data.

Reply:
> This means that a query that is only a selector will have a total of 0, as no math was done.

Only when I remove the current line.

Reply:
Any suggestions here? @brian-brazil

Reply:
>  A rate(1h) with a 1m step is more expensive than a rate 1h with a 1h step over the same data.

In that case, the current implementation does not seem wrong as those samples are being touched that are responsible for making queries expensive. We should consider again even though they have been considered during VS since they are being calculated again in the loops which increase the expense. Otherwise, if all samples other than the ones considered at VS are ignored, meaning, we should only consider the ones at VS which is pointless.

Reply:
That's still double counting, we shouldn't count them here - only when we do math on them subsequently.

Reply:
Is there any other way in this then?

Reply:
Ahh, I see you mean that not to count here as 0 is acceptable to us.

Reply:
So, is 0 samples, in this case, is acceptable to us?

Reply:
Yeah.

We might add on the output samples at the very end just so it doesn't look too weird to users.

Reply:
> We might add on the output samples at the very end just so it doesn't look too weird to users.

Assuming that will be in this PR, should I keep it as a different property as `outputSamples` alongside with `totalSamples` and `peakSamples`?

Reply:
I meant add it into totalSamples.

There's no need for an outputSamples, the user can already calculate that from the output.